### PR TITLE
Fix compatibility with upcoming cocotb 2.0

### PR DESCRIPTION
**Pull request by p12tic**
This PR fixes remaining issues that prevented compatibility with cocotb 2.0, mostly BinaryValue -> LogicArray migration.

CI for cocotb 2.0 has been enabled and compatibility with cocotb 2.0 has been declared in setup.py.

Note that latest cocotb master doesn't work due to https://github.com/cocotb/cocotb/commit/bfd811bbeba4c7dcc045beeb5c78f8e4901c8364 breaking something.

### CODE REVIEW

**Comment by p12tic** [comment]
@cmarqu @ktbarrett Just a friendly ping..

**Comment by ktbarrett** [comment]
I'll look at this later today.

File: .github/workflows/tests.yml
Comment (ktbarrett) [inline .github/workflows/tests.yml:29]: Can we just use `"master"` here?

Comment (p12tic) [reply .github/workflows/tests.yml:29]: https://github.com/cocotb/cocotb/commit/bfd811bbeba4c7dcc045beeb5c78f8e4901c8364 breaks something, thus we cannot use master (at least as of 2 weeks ago).

My plan was to get this PR through review and then, once the code is verified to be correct, raise a bug on cocotb repository.

Comment (p12tic) [reply .github/workflows/tests.yml:29]: Other than that, I would actually recommend against using simple `master`, because this makes tests unstable. A future commit may break cocotb-bus CI. Then whoever makes a PR will be forced to fix this issue as well. Better to pin the dependencies and fix issues at our own speed.

Comment (ktbarrett) [reply .github/workflows/tests.yml:29]: Alright. The break is that setimmediate works as intended now. All setimmediate in cocotb-bus should be updated to regular value setting since that's what it was doing before.

File: src/cocotb_bus/compat.py
Comment (ktbarrett) [inline src/cocotb_bus/compat.py:67]: This doesn't seem useful. `str()` casts work on both, right?

File: examples/endian_swapper/tests/test_endian_swapper.py
Comment (ktbarrett) [inline examples/endian_swapper/tests/test_endian_swapper.py:226]: Is there a reason to not just use `.integer` here if the intent is to support both 1.X and 2.X?

File: setup.py
Comment (p12tic) [reply setup.py:34]: Applied, thanks.

File: src/cocotb_bus/compat.py
Comment (p12tic) [reply src/cocotb_bus/compat.py]: Indeed, fixed now. Thanks.

File: examples/endian_swapper/tests/test_endian_swapper.py
Comment (p12tic) [reply examples/endian_swapper/tests/test_endian_swapper.py:226]: The problem is that `.integer` throws a deprecation warning on 2.0 which is hard to turn off. It wouldn't be a problem if cocotb-bus was a leaf package, then we could turn off the warning globally and call it a day. However, cocotb-bus can be used by third-party packages as well. This means that we should disable the warning on each call site. Since this functionality would be very similar in each call site, we would be tempted to extract it to a function and thus we would arrive to something like `convert_binary_to_unsigned` again.

File: .github/workflows/tests.yml
Comment (p12tic) [reply .github/workflows/tests.yml]: Fixed now and also upgraded cocotb to latest master.

**Comment by p12tic** [comment]
@ktbarrett The PR is ready for a review again.

**Review by ktbarrett — COMMENTED** [review]
I can't say I love the approach of the compat layer, but nothing else comes to mind.

I was considering dropping the DeprecationWarnings on the LogicArray interfaces until 2.1, so there's at least one release where there's no warning, but then we have the same issue once 2.1 drops, so that doesn't solve anything really. I guess this is just proof its easier to get it right the first time.

**Review by ktbarrett — APPROVED** [review]

**Comment by p12tic** [comment]
> I was considering dropping the DeprecationWarnings on the LogicArray interfaces until 2.1, so there's at least one release where there's no warning, but then we have the same issue once 2.1 drops, so that doesn't solve anything really. I guess this is just proof its easier to get it right the first time.

The root cause is that cocotb-bus supports cocotb as old as 1.6. If only 1.9.x and 2.x are supported then it's possible to backport a lot of things to 1.9.x and reduce the amount of deprecation warnings without a solution. E.g. missing functions can be added to BinaryValue so that similar semantics are supported and a switch from BinaryValue to LogicArray changes nothing at the source level.

**Comment by ktbarrett** [comment]
@p12tic I think this PR can come in as is, but that does bring up a good point. We only ever support the most recently released version of cocotb, so removing support for cocotb<1.9, and only supporting 1.9 and 2.0+ from here on out is totally fine by us. Especially considering that I'm pretty sure the only changes that have been made since the last release are pretty minor, so I don't see doing so as gatekeeping new features.

**Comment by p12tic** [comment]
>  Should we still do a release of the current state which supports older cocotb versions?

The PR still supports the older cocotb versions, so there's need for new release from before this PR.

On the other hand, this PR drops requirement on `cocotb<2.0`, so releasing after merging makes sense as it blocks people from using cocotb master branch, which has a few improvements unrelated to python API, e.g. simulator improvements and so on.

**Comment by ktbarrett** [comment]
@p12tic Sorry about the delay. We usually like to have more than one maintainer approve larger PRs or merge them after a couple days if there are no comments, but this got lost.

**Pull request merged**
